### DISCUSSION TITLE
Always-On & On-Demand VPN

### DISCUSSION THREAD

**Discussion by ChrisCrossQS** [strongswan/strongswan#2313] [Q&A]
First of all my deep thank you and appreciation and thanks to the Dev-Team. The new EMM Appconfig possibilities are awesome!

Background:
In the past I used Samsung OEMconfig in order to deploy a VPN profile to the Android Clients. It works well in Profile and Device Owner Devices but it does not work in a setup where you are Device Owner but still want to have a Profile Owner Separation (clarified this with Samsung Support - stopped working since Version "S"). Google now calls this setup work-profile on a company owned device (w-PC). For reference the link to the Knox VPN documentation: https://docs.samsungknox.com/admin/fundamentals/whitepaper/samsung-knox-for-android/network-security/virtual-private-networks/
The Samsung built-in VPN seems to be based on StrongSwan (I referes to it as built-in StrongSwan), thus I tested it.
I deployed the StrongSwan App in the work profile on a w-PC Device, configured it to work with our StrongSwan Server and it works well. I even got always-on VPN working (I was surprised the VPN conenction is visible in the settings in the personal profile although the VPN is setup in the work profile) The only thing I'm missing now is the on-demenad feature, I used to love in Samsung Knox VPN.

My Questions:
Is there a possibility in the Android App to configure an "on-demand" VPN, meaning the VPN is established only when the configured apps want to communicate with the server (e.g. email app with email server)? This would mean the VPN would not be established all the time but just when necessary.

**Comment by tobiasbrunner** [comment]
> Is there a possibility in the Android App to configure an "on-demand" VPN, meaning the VPN is established only when the configured apps want to communicate with the server (e.g. email app with email server)? 

I'm not aware of the VpnService API (or anything related to it) providing that possibility. Maybe there is a system-level option? (The Always-On feature is also only available there, the app itself has no control over it.)

It might be possible to script this via Intents (when or before an app is launched, start the VPN), or just enable Always-On VPN with a profile that only has specific apps enabled in the advanced settings.
